---
title: "Simple skeleton simulation"
author: "Martin Johnsson"
output: html_document
---

Simple simulation with these features:

* Discrete generations

* 3000 females, 200 males used each generation

* Selection only on the male side

* 6000 matings with equal sex ratio => 200 out of 3000 (3.3% of males used)

* One fully recessive lethal ~ 5% starting frequency

* No carrier x carrier matings; carrier x noncarrier matings allowed

## Housekeeping

Load packages and simulation results

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
##knitr::opts_knit$set(root.dir = normalizePath(".."))
```

```{r}
library(AlphaSimR)
library(dplyr)
library(ggplot2)
library(purrr)
library(tidyr)

source("R/simulation_functions.R")
```

```{r}
##simulation_results <- readRDS("simulations/simple_simulations/results_2.Rds")
simulation_results <- lapply(paste("simulations/simple_simulations/results_",
                                   1:10,
                                   ".Rds",
                                   sep = ""),
                             readRDS)
names(simulation_results) <- 1:10
```

Generation counter

```{r}
generations <- 1:40
```


## Basic features of the simulation

Genetic trend

```{r}
stats <- map_df(simulation_results, function(x) x$stats)
stats$replicate <- rep(1:10, each = length(generations))

qplot(x = generation,
      y = mean_g,
      colour = replicate,
      data = stats,
      xlab = "Generation",
      ylab = "Mean genetic value",
      main = "Genetic trend")
```

Genetic variance

```{r}
qplot(x = generation,
      y = var_g,
      data = stats,
      colour = replicate,
      xlab = "Generation",
      ylab = "Genetic variance",
      main = "Genetic variance trend")
```

## Allele frequency changes

Lethal allele frequency

```{r}
carriers <- map_df(simulation_results, function(x) x$carriers)
carriers$replicate <- rep(1:10, each = length(generations))

qplot(x = generation,
      y = carriers/6400,
      data = carriers,
      colour = replicate,
      xlab = "Generation",
      ylab = "Lethal allele frequency")
```


Frequencies of neutral variants

Example with replicate 1

```{r}
generations <- readRDS("simulations/simple_simulations/populations_1.Rds")
counts <- lapply(generations,
                     function(gen) {
                          snps <- pullSnpGeno(gen,
                                              simParam = simulation_results[[1]]$simparam)
                          snps <- snps[, simulation_results[[1]]$other_snp_ix]
                          colSums(snps)
                      })
counts_df <- as.data.frame(Reduce(rbind, counts))
counts_df$generation <- 1:40
counts_long <- pivot_longer(counts_df,
                            -generation)
qplot(x = generation,
      y = value/6400,
      group = name,
      data = counts_long,
      geom = "line",
      xlab = "Generation",
      ylab = "Allele frequency",
      main = "Allele frequency of neutral alleles with ~ 5% starting frequency")
```

Allele frequency summary

```{r}
freq_summaries <- summarise(group_by(counts_long, generation),
                            mean = mean(value/6400),
                            lower = mean - 2 * sd(value/6400),
                            upper = mean - 2 * sd(value/6400))
```


Combined plot

```{r}

ggplot() +
    geom_line(aes(x = generation,
                  y = value/6400,
                  group = name),
              data = counts_long,
              colour = "grey") +
    geom_line(aes(x = generation,
                  y = carriers/6400,
                  colour = replicate,
                  group = replicate),
              data = carriers) +
    xlab("Generations") +
    ylab("Allele frequency") +
    ggtitle("Allele frequency of lethal (blue) and neutral alleles (grey)")
```

