---
title: "Simple skeleton simulation"
author: "Martin Johnsson"
output: html_document
---


## Housekeeping

Load packages and simulation results

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
##knitr::opts_knit$set(root.dir = normalizePath(".."))
```

```{r}
library(AlphaSimR)
library(dplyr)
library(ggplot2)
library(purrr)
library(tidyr)

source("R/simulation_functions.R")
```

```{r}
filenames <- system("ls simulations/simple_simulations/avoid_carrier_x_carrier_snp/results_*.Rds",
                    intern = TRUE)
simulation_results <- read_results(filenames)
```

Generation counter

```{r}
generations <- 1:20
```



## Basic features of the simulation

Genetic trend

```{r}
stats <- combined_stats(simulation_results)

(genetic_trend_plot(stats))
```

Genetic variance

```{r}
(genetic_variance_plot(stats))
```

## Allele frequency changes

### Lethal allele frequency

```{r}
carriers <- combined_carriers(simulation_results)

lethal_frequency_plot(carriers) +
    ggtitle("Lethal allele frequency change without balancing selection")
```


```{r}
end_carriers <- get_end_carriers(carriers)
mean(end_carriers$average_carriers/2/end_carriers$average_n)
sd(end_carriers$average_carriers/2/end_carriers$average_n)
```



### Comparison of lethals with neutral variants

Example with replicate 1

```{r}
populations <- readRDS("simulations/simple_simulations/avoid_carrier_x_carrier_snp/populations_1.Rds")
counts <- lapply(populations,
                     function(gen) {
                          snps <- pullSnpGeno(gen,
                                              simParam = simulation_results[[1]]$simparam)
                          snps <- snps[, simulation_results[[1]]$other_snp_ix]
                          colSums(snps)
                      })
counts_df <- as.data.frame(Reduce(rbind, counts))
counts_df$generation <- generations
counts_long <- pivot_longer(counts_df,
                            -generation)
qplot(x = generation,
      y = value/2/6000,
      group = name,
      data = counts_long,
      geom = "line",
      xlab = "Generation",
      ylab = "Allele frequency",
      main = "Allele frequency of neutral alleles with ~ 5% starting frequency")
```



Combined plot

```{r}

ggplot() +
    geom_line(aes(x = generation,
                  y = value/2/6000,
                  group = name),
              data = counts_long,
              colour = "grey") +
    geom_line(aes(x = generation,
                  y = carriers/2/6000,
                  colour = replicate,
                  group = replicate),
              data = carriers) +
    xlab("Generations") +
    ylab("Allele frequency") +
    ggtitle("Allele frequency of lethal (blue) and neutral alleles (grey)")
```

